src/probnum/linalg/solvers/_state.py

```diff
@@ -0,0 +1,126 @@
+
+    The solver state contains the problem to be solved, the current belief over the quantities of interest and any miscellaneous quantities computed during an iteration
+    of a probabilistic linear solver. The solver state is passed between the
+    different components of the solver and may be used internally to cache quantities which are used more than once.
+
+    Parameters
+    ----------
+    problem
+        Linear system to be solved.
+    prior
+        Prior belief over the quantities of interest of the linear system.
+    rng
+        Random number generator.
+    """
+
+    def __init__(
+        self,
+        problem: problems.LinearSystem,
+        prior: "probnum.linalg.solvers.beliefs.LinearSystemBelief",
+        rng: Optional[np.random.Generator] = None,
+    ):
+        self.problem = problem
+        self.belief = prior
+
+        self.step = 0
+
+        self._actions: List[np.ndarray] = [None]
+        self._observations: List[Any] = [None]
+        self._residuals: List[np.ndarray] = [
+            self.problem.A @ self.belief.x.mean - self.problem.b,
+            None,
+        ]
+
+        self._cache = defaultdict(list)
+        self.rng = rng
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}(step={self.step})"
+
+    @property
+    def action(self) -> Optional[np.ndarray]:
+        """Action of the solver for the current step.
+
+        Is ``None`` at the beginning of a step and will be set by the
+        policy.
+        """
+        return self._actions[self.step]
+
+    @action.setter
+    def action(self, value: np.ndarray) -> None:
+        assert self._actions[self.step] is None
+        self._actions[self.step] = value
+
+    @property
+    def observation(self) -> Optional[Any]:
+        """Observation of the solver for the current step.
+
+        Is ``None`` at the beginning of a step, will be set by the
+        observation model for a given action.
+        """
+        return self._observations[self.step]
+
+    @observation.setter
+    def observation(self, value: Any) -> None:
+        assert self._observations[self.step] is None
+        self._observations[self.step] = value
+
+    @property
+    def actions(self) -> Tuple[np.ndarray, ...]:
+        """Actions taken by the solver."""
+        return tuple(self._actions)
+
+    @property
+    def observations(self) -> Tuple[Any, ...]:
+        """Observations of the problem by the solver."""
+        return tuple(self._observations)
+
+    @property
+    def residual(self) -> np.ndarray:
+        r"""Residual :math:`Ax-b` for the current step."""
+        if self._residuals[self.step] is None:
+            self._residuals[self.step] = (
+                self.problem.A @ self.belief.x.mean - self.problem.b
+            )
+        return self._residuals[self.step]
+
+    @property
+    def residuals(self) -> Tuple[np.ndarray, ...]:
+        r"""Residuals :math:`\{Ax_i - b\}_i`."""
+        return tuple(self._residuals)
```

Comment:
Currently this would mean some entries in `residuals` will be `None` as for `actions` and `observations`. I don't see this as a problem. Am I missing something?

Reply:
This might be nasty for an outside user, since it is not really transparent when the residuals are `None` and when they are `np.ndarray`.
I think it's fine for now though.